g15helper/g15helper_emu.cpp

```diff
@@ -0,0 +1,125 @@
+// Copyright 2005-2016 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+/*
+ * G15 Helper Emulator
+ */
+
+#include <cstdio>
+#include <windows.h>
+
+#include <QApplication>
+
+#include "g15helper.h"
+#include "g15helper_emu.h"
+
+#if defined(Q_OS_WIN)
+# include <io.h>
+# include <fcntl.h>
+#endif
+
+G15Reader::G15Reader()
+	: QThread() {
+
+	moveToThread(this);
+
+	m_isRunning = true;
+}
+
+G15Reader::~G15Reader() {
+	m_isRunning = false;
+}
+
+void G15Reader::run() {
+#if defined(Q_OS_WIN)
+	_setmode(_fileno(stdin), O_BINARY);
+#endif
+
+	quint8 buf[G15_MAX_FBMEM];
+
+	QImage img(QSize(G15_MAX_WIDTH, G15_MAX_HEIGHT), QImage::Format_MonoLSB);
+
+	while (m_isRunning) {
+		quint8 priority = 0;
+		size_t nread = 0;
+		size_t ntotal = 0;
+
+		memset(&buf, 0, G15_MAX_FBMEM);
+
+		nread = fread(&priority, 1, sizeof(priority), stdin);
+		if (nread <= 0) {
+			qFatal("g15helper: unable to read stdin, retval %lli",
+			       static_cast<long long>(nread));
+		}
+
+		do {
+			nread = fread(&buf[0] + ntotal, 1, G15_MAX_FBMEM - ntotal, stdin);
+			if (nread <= 0) {
+				qFatal("g15helper: unable to read stdin, retval %lli",
+				       static_cast<long long>(nread));
+			}
+			ntotal += nread;
+		} while (ntotal < G15_MAX_FBMEM);
+
+		for (int w = 0; w < G15_MAX_WIDTH; w++) {
+			for (int h = 0; h < G15_MAX_HEIGHT; h++) {
+				quint8 color = buf[G15_MAX_WIDTH * h + w];
+				uint val = color == 0xff ? 1: 0;
```

Comment:
Is this really strictly mono for all g15 displays? Otherwise ```QImage::Format_RGB888``` for grayscale wouldn't really add any complexity from what I can see.